programs/lib/thread.c

```diff
@@ -0,0 +1,98 @@
+#include <stdbool.h>
+#include <flexpret_io.h>
+#include <flexpret_lock.h>
+#include <flexpret_thread.h>
+
+/* Arrays that keep track of the status of threads */
+static void* (*routines[NUM_THREADS])(void *); // An array of function pointers
+static void** args[NUM_THREADS];
+static bool exit_requested[NUM_THREADS];
+static void** exit_code[NUM_THREADS];
+static bool in_use[NUM_THREADS]; // Whether a thread is currently executing a routine.
+
+// Keep track of the number of threads
+// currently processing routines.
+//
+// Hardware threads being not "busy"
+// (i.e., running a routine) is not
+// enough to terminate the main thread.
+// To terminate the main thread,
+// all hardware worker threads need
+// to be EXITED.
+uint32_t num_threads_busy = 0;
+
+// Keep track of the number of threads
+// currently processing routines.
+uint32_t num_threads_exited = 0;
+
+// More precisely, this is a "thread_map",
+// i.e. mapping a routine to a running
+// hardware thread.
+int thread_create(
+    thread_t *restrict hartid,
+    void *(*start_routine)(void *),
+    void *restrict arg
+) {
+    // Allocate an available thread.
+    // Cannot allocate to thread 0.
+    for (int i = 1; i < NUM_THREADS; i++) {
+        if (!in_use[i]) {
```

Comment:
There might be a race condition here? If multiple threads call `thread_create` at the same time. I think the safest is to acquire the lock at the beginning of the function call

Reply:
Yes, this is a good call. Fixing it now.